2단계 - 로또(자동)

h2jinee opened the pull request:
## 📌 Summary
- 로또 자동 구매 및 당첨 통계 기능을 구현했습니다. (PR 제출하고 3단계 수강신청을 진행해보겠습니다!)
- 1단계 문자열 계산기 피드백을 반영했습니다.
- MVC 패턴, 책임과 역할을 고민하며 클래스를 분리했습니다.
- 테스트 코드 작성 시 `@DisplayName`를 사용하지 않게 됐습니다.. ㅋㅋ 메서드명을 영어로 작성하니 이름 짓는 데 고민하는 시간이 많이 소요돼서 비효율적이라고 느꼈습니다. 😂
- 객체 지향의 사실과 오해 책을 같이 읽으면서 미션을 진행하고 있는데요! 절반정도 읽었는데, 미션을 진행하면서 같이 보니 시야가 조금씩 넓어지고 있는 게 체감이 되고 있습니다.. 🥹

## 💬 Review Points
- **관심사 분리**:
    - Lotto: 로또 번호 6개 보관 및 검증, 번호 일치 개수 계산
    - LottoPurchase: 구매 금액 검증 및 구매 개수 계산
    - LottoRank (enum): 등수별 일치 개수/상금 매핑
    - LottoStatistics: 당첨 통계 및 수익률 계산
    - LottoGenerator: 로또 번호 생성
    - InputView / OutputView: 입출력 처리
- 처음에 `Lotto`가 구매 금액을 담고 있도록 접근을 했습니다, 이후 `WinningLotto`를 만들면서 타입 불일치 문제가 발생하여 다시 책임과 역할을 다시 정리해서 분리했습니다. Lotto는 본질에 맞게 "로또 번호 6개"를 담을 수 있도록 수정하였고, 구매 금액은 `LottoPurchase`로 분리했습니다.

## 🤔 고민한 점 및 질문
- **검증 로직 분리**:
    - validate를 구현할 때마다 고민이 되었는데요.. 여러 검증을 하나의 메서드에 다 넣는 게 맞는 건가요?
- **방어적 복사와 불변**:
    - 테스트 작성 중 `List.of()`로 생성한 로또에 `Collections.sort(numbers);`를 사용해서 정렬하려다 `UnsupportedOperationException`이 발생했습니다.
    - 원인을 해결하던 중 AI를 통해 원인이 방어적 복사를 하지 않아서임을 알게 되었는데, 방어적 복사라는 개념이 잘 이해되지 않습니다..
- **enum 메서드 네이밍 충돌**:
    - `LottoRank`에 `valueOf(int matchCount)` 메서드를 만들었는데, enum 기본 제공 `valueOf(String)`과 충돌해서 메서드 레퍼런스(`.map(LottoRank::valueOf)`)가 되지 않았습니다. AI를 통해 원인을 파악해 `fromMatchCount()`로 이름을 바꿔서 해결했습니다.
    - `of`, `from`, `valueOf` 등 이런 메서드에 대해서 완벽하게 이해를 하지 못했는데, 이 부분은 좀 더 학습이 필요하다고 느꼈습니다.

h2jinee committed:
1단계 피드백 반영

h2jinee committed:
feat: 로또 기능 추가 및 입력/출력 뷰 리팩토링

- 로또 번호 입력 기능과 로또 구매 기능을 추가했습니다.
- InputView와 OutputView 클래스를 lotto.view 패키지로 이동했습니다.
- CalculatorMain에서 InputView, OutputView를 calculator.view로 수정했습니다.

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+public record Lotto(List<Integer> numbers) {
+    public static final int LOTTO_COUNT = 6;
+    public static final int START_NUMBER = 1;
+    public static final int END_NUMBER = 45;
```
```suggestion
    public static final int MIN_NUMBER = 1;
    public static final int MAX_NUMBER = 45;
```
이름은 최소/최대가 더 적합할 듯

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+    public Lotto {
+        numbers = new ArrayList<>(numbers);
+        Collections.sort(numbers);
+        validate(numbers);
```
유효성 체크를 먼저하는 것이 낫지 않을까?

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+                .count();
+    }
+
+    static void validate(List<Integer> list) {
```
유효성 체크에 대한 정답은 없는지라..
복잡도 증가한다면 유효성 체크를 더 간단히 할 수 있거나, 다른 객체로 분리해서 할 수는 없을까 고민해 보는 것도 방법이지 않을까?

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+            throw new IllegalArgumentException("로또 번호는 6개여야 합니다.");
+        }
+
+        if (list.stream().distinct().count() != list.size()) {
```
Lotto의 6개 값은 중복을 허용하지 않는다는 특성을 반영하기 위해 Lotto 필드인 List를 Set로 구현해 보면 어떨까?
Set으로 관리하다 외부에서 이 값을 접근할 때 순서를 보장해야 한다면 정렬해서 반환하는 것은 어떨까?

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,45 @@
+            if (number < START_NUMBER || number > END_NUMBER) {
+                throw new IllegalArgumentException("로또 번호는 1부터 45 사이의 숫자여야 합니다.");
+            }
+        });
```
이 유효성 체크는 다른 객체에 위임할 수는 없을까?
예를 들어 List 콜렉션의 값이 integer 원시값을 포장한다면...

javajigi commented on src/main/java/lotto/LottoRank.java:
```diff
@@ -0,0 +1,33 @@
+        return Arrays.stream(LottoRank.values())
+                .filter(rank -> rank.matchCount == matchCount)
+                .findFirst()
+                .orElse(null);
```
null을 반환하기 보다 꽝을 의미하는 enum을 추가해 반환한다면...
null object 패턴을 검색해 학습하고 적용해 본다.

javajigi commented on src/main/java/lotto/LottoMain.java:
```diff
@@ -0,0 +1,30 @@
+        Lotto winningLotto = InputView.inputWinningNumbers();
+
+        // 당첨 통계 및 수익률 출력
+        Map<LottoRank, Integer> statistics = calculateStatistics(lotto, winningLotto);
```
```suggestion
        LottoStatistics statistics = calculateStatistics(lotto, winningLotto);
```
Map을 포장한 일급 콜렉션이 있는데 이 객체를 적극적으로 사용해 보면 어떨까?

h2jinee replied:
피드백 감사합니다! 말씀해주신 대로 `LottoStatistics` 타입으로 받도록 수정하려고 보니 기존에는 `static` 메서드를 사용해서 `Map`을 파라미터로 넘기는 방식이었는데, `LottoStatistics`를 쓰려면 결국 내부 Map을 다시 꺼내야 하더라고요.. 찾아보니 `record`는 자동으로 `public getter`를 만들어서 `Map`이 외부에 노출되는 문제가 있다는 것을 알게 되었습니다.. 그래서 `Map.copyOf()`로 불변화하고, 인스턴스 메서드로 바꿔서 `statistics.calculateYield(amount)` 형태로 객체에게 직접 일을 시키도록 변경했습니다. 일급 컬렉션을 왜 강조하시는지 알 것 같아요.. 감사합니다! 🥹

javajigi commented on src/test/java/lotto/LottoRankTest.java:
```diff
@@ -0,0 +1,46 @@
+    
+    @Test
+    void 당첨_통계에서_각_등수별_당첨_개수를_셀_수_있다() {
+        Lotto winningLotto = new Lotto(List.of(1, 2, 3, 4, 5, 6));
```
```suggestion
        Lotto winningLotto = new Lotto(1, 2, 3, 4, 5, 6);
```
가변 인자를 가지는 생성자를 추가하면 위와 같이 더 편리하게 구현할 수 있지 않을까?
다수의 부 생성자를 추가해 개발자의 편리성을 제공해 본다.

javajigi commented on src/test/java/lotto/LottoRankTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertThat(rankCount.get(FIRST)).isEqualTo(1L);
+        assertThat(rankCount.get(SECOND)).isEqualTo(1L);
+        assertThat(rankCount.get(THIRD)).isEqualTo(1L);
+        assertThat(rankCount.get(FOURTH)).isEqualTo(1L);
```
로또 실행 결과를 LottoStatistics에 저장하고 있는데 이 단위테스트는 LottoStatistics에 대한 단위 테스트로 추가해야하지 않을까?

javajigi commented on src/main/java/lotto/LottoStatistics.java:
```diff
@@ -0,0 +1,31 @@
+
+public record LottoStatistics(Map<LottoRank, Integer> statistics) {
+
+    static Map<LottoRank, Integer> calculateStatistics(List<Lotto> lottoList, Lotto winningLotto) {
```
List<Lotto> lottoList을 일급 콜렉션으로 구현해 본다.
이렇게 일급 콜렉션 객체를 추가했을 때 이 메서드 로직 구현의 책임을 어느 객체에 부여할 것인지 고민해 본다.

javajigi reviewed (changes requested):
로또 미션 구현 시작했군요. 👍 
전체적인 객체 설계 나쁘지 않네요. 그럼에도 불구하고 약간의 아쉬움이 남는 부분이 있어 피드백 남겨봅니다.
특히 자동차 경주 미션에서 적용했던 원시값 포장, 일급 콜렉션 객체 분리를 좀 더 적극적으로 해볼 것을 추천합니다.

h2jinee committed:
refactor: 2단계 피드백 반영

- LottoNumber: 원시값을 포장하여 1~45 범위 검증 책임을 위임했습니다
- Lottos: List<Lotto>를 포장한 일급 컬렉션을 구현했습니다
- LottoStatistics: Map.copyOf()로 불변화하고, 인스턴스 메서드로 리팩토링했습니다
- LottoRank: NO_RANK 추가로 Null Object 패턴을 적용했습니다
- Lotto: 가변 인자 부 생성자를 추가하였습니다
- OutputView: 캡슐화를 준수하도록 메서드 호출 방식을 변경했습니다
- 테스트 코드를 리팩토링하고 테스트 위치를 정리했습니다

javajigi commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,16 @@
+package lotto;
+
+public record LottoNumber(int value) {
```
👍

javajigi commented on src/main/java/lotto/LottoStatistics.java:
```diff
@@ -0,0 +1,41 @@
+        Map<LottoRank, Integer> result = new HashMap<>();
+
+        for (Lotto lotto : lottoList.lottos()) {
+            int matchCount = lotto.matchCount(winningLotto);
+            LottoRank rank = LottoRank.fromMatchCount(matchCount);
```
```suggestion
            LottoRank rank = lotto.rank(winningLotto);
```
위와 같이 구현하는 것은 어떨까?

javajigi commented on src/test/java/lotto/LottoStatisticsTest.java:
```diff
@@ -0,0 +1,23 @@
+                new Lotto(1, 2, 3, 4, 5, 7),
+                new Lotto(1, 2, 3, 4, 8, 9)
+        );
+        Lotto winningLotto = new Lotto(1, 2, 3, 4, 5, 6);
```
가변인자를 가지는 생성자 추가 후 구현 👍

javajigi reviewed (approved):
피드백 반영 넘 잘 했네요. 💯 
3단계 도전할 때 객체의 변경이 얼마나 많이 발생하는지 느껴보면 좋겠습니다.
현재 객체 설계가 잘 되어 있다면 변경을 최소화하면서 기능 추가가 가능할 겁니다.
